realm/realm-library/src/main/java/io/realm/RealmConfiguration.java

```diff
@@ -615,16 +633,36 @@ public Builder initialData(Realm.Transaction transaction) {
          * When opening the Realm for the first time, instead of creating an empty file,
          * the Realm file will be copied from the provided asset file and used instead.
          * <p>
-         * <p>This cannot be configured to clear and recreate schema by calling {@link #deleteRealmIfMigrationNeeded()}
-         * at the same time as doing so will delete the copied asset schema.
-         * <p>
+         * This cannot be combined with {@link #deleteRealmIfMigrationNeeded()} as doing so would just result in the
+         * copied file being deleted.
          * <p>
          * WARNING: This could potentially be a lengthy operation and should ideally be done on a background thread.
          *
          * @param assetFile path to the asset database file.
          * @throws IllegalStateException if this is configured to clear its schema by calling {@link #deleteRealmIfMigrationNeeded()}.
          */
         public Builder assetFile(final String assetFile) {
+            return assetFile(assetFile, false);
+        }
+
+        /**
+         * Copies the Realm file from the given asset file path.
+         * <p>
+         * When opening the Realm for the first time, instead of creating an empty file,
+         * the Realm file will be copied from the provided asset file and used instead.
+         * <p>
+         * This cannot be combined with {@link #deleteRealmIfMigrationNeeded()} as doing so would just result in the
+         * copied file being deleted.
+         * <p>
+         * WARNING: This could potentially be a lengthy operation and should ideally be done on a background thread.
+         *
+         * @param assetFile path to the asset database file.
+         * @param readOnly if {@code true} no local changes can be made to the Realm once copied. All write
+         * transactions will fail with an {@link IllegalStateException}. Marking a Realm as read-only only applies to
+         * the Realm in this process. Other processes can still write to the Realm.
+         * @throws IllegalStateException if this is configured to clear its schema by calling {@link #deleteRealmIfMigrationNeeded()}.
+         */
+        public Builder assetFile(final String assetFile, boolean readOnly) {
```

Comment:
In the original design `readOnly()` was a separate builder method, but I don't see it being useful unless you have an assetFile? I would actually argue that setting `readOnly()` in all other cases is probably a programmer error.
It would be more future proof though if we decide to allow it for other use cases and it removes the ugliness that is a un-named boolean on `waitForInitialRemoteData(false)` which just reads wrong on many levels. 

I briefly considered allowing it for `initialTransaction()` as well, but that will definitely conflict with the current semantics from Object Store, but we might want to change that when we migrate?

I won't be opposed to add `Builder.readOnly()` as it massively increases readability, but at an increased chance of people using it wrongly. Thoughts? @realm/java